tokio/src/sync/broadcast.rs

```diff
@@ -640,15 +646,21 @@ impl<T> Drop for Sender<T> {
 
 impl<T> Receiver<T> {
     /// Locks the next value if there is one.
-    fn recv_ref(&mut self) -> Result<RecvGuard<'_, T>, TryRecvError> {
+    fn recv_ref(
+        &mut self,
+        waiter: Option<(&UnsafeCell<Waiter>, &Waker)>,
```

Comment:
This seems like it should maybe take a pinned waiter? We would still have to use `Pin::get_unchecked_mut` internally to turn the pin into a `NonNull`, but it would at least make it a little clearer to callers that the waiter has to be pinned. Otherwise, this should probably be unsafe to call?

Reply:
The waiter should be pinned, but it still needs `UnsafeCell`. As this is all a module internal detail, IMO this falls in the case of being better to not define this fn as `unsafe` in order to not allow arbitrary unsafe usage inside the fn.

Reply:
Hmm...could we add a comment, at least?

As far as I can tell, everywhere this is called with a waiter (and not a `None`), the waiter already is pinned anyway? I think the only effect of changing this to take `Pin<&mut UnsafeCell<Waiter>>` would be that any _new_ code that calls this method with a waiter that isn't already pinned would need to add `unsafe` at the callsite — everywhere we call this currently, the waiter already is pinned.